spotify.go

```diff
@@ -107,14 +118,50 @@ func (e Error) Error() string {
 }
 
 // decodeError decodes an Error from an io.Reader.
-func decodeError(r io.Reader) error {
+func decodeError(c *Client, resp *http.Response) error {
+	responseBody, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	if len(responseBody) == 0 {
+		return fmt.Errorf("spotify: there was an HTTP error (%d '%s') but the body was empty", resp.StatusCode, http.StatusText(resp.StatusCode))
+	}
+
+	buf := bytes.NewBuffer(responseBody)
+
 	var e struct {
 		E Error `json:"error"`
 	}
-	err := json.NewDecoder(r).Decode(&e)
+	err = json.NewDecoder(buf).Decode(&e)
 	if err != nil {
-		return errors.New("spotify: couldn't decode error")
+		return fmt.Errorf("spotify: couldn't decode error: (%d) [%s]", len(responseBody), responseBody)
+	}
+
+	if resp.StatusCode == rateLimitExceededStatusCode {
+		retrySecondsRaw := resp.Header.Get("Retry-After")
+		if retrySecondsRaw != "" {
+			retrySeconds, err := strconv.ParseInt(retrySecondsRaw, 10, 32)
+			if err != nil {
+				return fmt.Errorf("could not parse retry seconds: %s", retrySecondsRaw)
+			}
+
+			if retrySeconds == 0 {
+				c.retryDuration = defaultRetryDuration
+			} else {
+				c.retryDuration = time.Second * time.Duration(retrySeconds)
+			}
+		}
+	} else if e.E.Error() == "" {
+		// Some errors will result in there being a useful status-code but an
+		// empty message, which will confuse the user (who only has access to
+		// the message and not the code). An example of this is when we send
+		// some of the arguments directly in the HTTP query and the URL ends-up
+		// being too long.
+
+		e.E.Message = fmt.Sprintf("spotify: there was an HTTP error (%d '%s') but we received an empty error", resp.StatusCode, http.StatusText(resp.StatusCode))
```

Comment:
Simplify this message too.  Something as simple as `spotify: unexpected HTTP %d: %s`